### PR TITLE
Decouple quarkus bootstrap from Maven implementation (#2251)

### PR DESCRIPTION
**Pull request by jskov**
This prepares the way for adding a new bootstrap implementation that
supports Gradle as a builder.

### CODE REVIEW

**Comment by jskov** [comment]
Sorry, I had clearly not checked for dependencies outside the module.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BuilderInfo.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.List;
+
+import io.quarkus.bootstrap.BootstrapDependencyProcessingException;
+import io.quarkus.bootstrap.BootstrapException;
+import io.quarkus.bootstrap.model.AppDependency;
+
+/**
+ * Provides access to information from builders such as Maven and Gradle.
+ */
+public interface BuilderInfo extends AutoCloseable {
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BuilderInfo.java]: I think we should keep the workspace API and discovery separate from the artifact and AppModel resolution. Perhaps we could split it in smaller tasks. E.g. have a Gradle LocalProject discovery mechanism. And then improve the existing AppModelGradleResolver.

Comment (jskov) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BuilderInfo.java]: I am not sure I understand what you mean.
Would you like BuilderInfos split into LocalProjects and ArtifactResolvers, with a new separate type ArtifactResolver that defines the resolution call?

Comment (aloubyansky) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BuilderInfo.java]: Yes, I wanted to separate the workspace discovery from the rest. There already is AppModelResolver interface and a limited Gradle build-time impl of it. I think we could take that as the base. What do you think?

**Comment by jskov** [comment]
I have moved the code a little around, hopefully addressing your concerns.

Cheers,
Jesper

**Comment by aloubyansky** [comment]
The original idea behind LocalWorkspace was just a workspace structure which consists of the projects info (their ID, sources, resources and output dirs). Classpath caching and artifact resolving APIs shouldn't belong there, imo. The purpose of ArtifactResolver isn't clear to me neither so far.

**Comment by jskov** [comment]
As I see it, you cannot have a LocalWorkspace that is not intimately bound to the builder which defines how that workspace is constructed.
From this follows that LocalProject's are also builder specific, and able to provide the (reference to) an implementation of the ArtifactResolver. Because this will be builder-specific.

The point of the ArtifactResolver interface, is simply to allow me to remove the 2-3 maven-specific lines you had at the bottom of the newDeploymentClassLoader method, replacing it with something that is builder neutral. The name comes from your MavenArtifactResolver class.

So I think the code makes makes sense in its current form.

But it is your work. You are welcome to take my suggested changes and rework them to something that you like better. 

Please let me know when you are done - then I will try to add the Gradle changes into whichever form the code ends up in. 

Cheers,
Jesper

**Pull request closed without merging**